tests/unit/test_config.py

```diff
@@ -276,3 +266,53 @@ def test_port_out_of_range(self, port):
             )
 
         assert "port out of range" in str(exc_info)
+
+
+class TestServiceUrlHelpers:
+    @pytest.mark.parametrize(
+        ["protocol", "subdomains", "host", "port", "expected_service_url"],
+        [
+            # Default
+            (None, None, None, None, "http://localhost.localstack.cloud:4566"),
+            # Customize each part with defaults
+            ("https", None, None, None, "https://localhost.localstack.cloud:4566"),
+            (None, "s3", None, None, "http://s3.localhost.localstack.cloud:4566"),
+            (None, None, "localstack-container", None, "http://localstack-container:4566"),
+            (None, None, None, 5555, "http://localhost.localstack.cloud:5555"),
+            # Multiple subdomains
+            (
+                None,
+                "abc123.execute-api.lambda",
+                None,
+                None,
+                "http://abc123.execute-api.lambda.localhost.localstack.cloud:4566",
+            ),
+            # Customize everything
+            (
+                "https",
+                "abc.execute-api",
+                "localstack-container",
+                5555,
+                "https://abc.execute-api.localstack-container:5555",
+            ),
+        ],
+    )
+    def test_external_service_url(
+        self,
+        protocol: str | None,
+        subdomains: str | None,
+        host: str | None,
+        port: int | None,
+        expected_service_url: str,
+    ):
+        url = external_service_url(host=host, port=port, protocol=protocol, subdomains=subdomains)
+        assert url == expected_service_url
+
+    def test_internal_service_url(self):
+        # defaults
+        assert internal_service_url() == "http://localhost:4566"
+        # subdomains
+        assert (
+            internal_service_url(subdomains="abc.execute-api")
+            == "http://abc.execute-api.localhost:4566"
+        )
```

Comment:
TIL that `localhost` [is a TLD](https://datatracker.ietf.org/doc/html/rfc2606#section-2)!